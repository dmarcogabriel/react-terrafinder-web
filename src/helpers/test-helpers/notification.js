export const mockUseNotification = (showNotification = () => {}) => {
  const mockShowNotification = jest.fn(showNotification);

  jest.mock('hooks/useNotification', () => ({
    useNotification: () => ({
      showNotification: mockShowNotification,
    }),
    NOTIFICATION_TYPES: { SUCCESS: 'success', ERROR: 'error' },
  }));

  return { mockShowNotification };
};
