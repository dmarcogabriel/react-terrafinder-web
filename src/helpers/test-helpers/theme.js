import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../styles/themes';

export const withTheme = (component) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component) =>
  renderer.create(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
