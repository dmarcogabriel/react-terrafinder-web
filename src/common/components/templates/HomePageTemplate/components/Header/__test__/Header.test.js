import React from 'react';
import ReactDOM from 'react-dom';
import { fireEvent, render } from '@testing-library/react';
import { renderWithTheme, withTheme } from 'helpers/test-helpers/theme';
import { Header } from '..';

let replaceResponse;
const mockReplace = jest.fn((url) => {
  replaceResponse = url;
});
let pushResponse;
const mockPush = jest.fn((url) => {
  pushResponse = url;
});

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    replace: mockReplace,
    push: mockPush,
  }),
  Link: ({ to, children }) => <a href={to}>{children}</a>,
}));

let mockUser;
const mockLogout = jest.fn();

jest.mock('hooks/useUser', () => ({
  useUser: () => ({
    currentUser: mockUser,
    userIsLogged: jest.fn(() => !!mockUser),
    logout: mockLogout,
  }),
}));

describe('common/components/templates/HomePageTemplate/Header', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(withTheme(<Header />), div);
    ReactDOM.unmountComponentAtNode(div);
  });

  describe('unlogged', () => {
    it('should handle click menu item', () => {
      const { getByTestId } = render(withTheme(<Header />));

      const menuButton = getByTestId('menuButton');
      fireEvent.click(menuButton);

      const firstMenuItem = getByTestId('menu-item-l1');
      fireEvent.click(firstMenuItem);

      expect(mockPush).toHaveBeenCalled();
      expect(pushResponse).toEqual('/search-property');
    });

    it('should handle onClose menu', () => {
      const { getByTestId } = render(withTheme(<Header />));

      const menuButton = getByTestId('menuButton');
      fireEvent.click(menuButton);
      fireEvent.click(menuButton);

      const menu = getByTestId('menu');
      expect(menu).toBeInTheDocument();
    });

    it('matches snapshot', () => {
      const tree = renderWithTheme(<Header />).toJSON('HeaderUnlogged');
      expect(tree).toMatchSnapshot();
    });
  });

  describe('logged', () => {
    it('should show grettings', () => {
      mockUser = { firstName: 'Tester' };
      const { getByTestId } = render(withTheme(<Header />));

      const greetings = getByTestId('greetings');
      expect(greetings).toBeInTheDocument();
      expect(greetings).toHaveTextContent('OlÃ¡, Tester');
    });

    it('should show logout button', () => {
      mockUser = { firstName: 'Tester' };
      const { getByTestId } = render(withTheme(<Header />));

      const logoutButton = getByTestId('logoutButton');
      expect(logoutButton).toBeInTheDocument();
      expect(logoutButton).toHaveTextContent('Logout');
    });

    it('should show menu', () => {
      const { getByTestId } = render(withTheme(<Header />));

      const menuButton = getByTestId('menuButton');
      fireEvent.click(menuButton);

      const menu = getByTestId('menu');
      expect(menu).toBeInTheDocument();
    });

    it('should handle logout', () => {
      mockUser = { firstName: 'Tester' };
      const { getByTestId } = render(withTheme(<Header />));

      const logoutButton = getByTestId('logoutButton');
      fireEvent.click(logoutButton);

      expect(mockLogout).toHaveBeenCalled();
      expect(mockReplace).toHaveBeenCalled();
      expect(replaceResponse).toEqual('/home');
    });

    it('matches snapshot', () => {
      mockUser = { firstName: 'Tester' };
      const tree = renderWithTheme(<Header />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
