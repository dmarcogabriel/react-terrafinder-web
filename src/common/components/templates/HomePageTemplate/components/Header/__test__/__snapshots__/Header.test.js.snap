// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/components/templates/HomePageTemplate/Header logged matches snapshot 1`] = `
.c0 {
  height: 4rem;
  cursor: pointer;
}

<header
  className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-zhueze-MuiPaper-root-MuiAppBar-root"
  data-testid="app-bar"
>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-nb5wfo-MuiToolbar-root"
  >
    <a
      href="/"
    >
      <img
        alt="logo"
        className="c0"
        src="\\"Logo.png\\""
      />
    </a>
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
      data-testid="menuButton"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="MenuIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-mzps2i-MuiToolbar-root"
  >
    <a
      href="/"
    >
      <img
        alt="logo desktop"
        className="c0"
        src="\\"Logo.svg\\""
      />
    </a>
    <a
      href="/dashboard"
    >
      Meus anúncios
    </a>
    <a
      href="/create/property"
    >
      Quero anunciar meu imóvel
    </a>
    <a
      href="/home/search-property"
    >
      Procurar Imóvel
    </a>
    <p
      className="MuiTypography-root MuiTypography-body1 css-jgi19-MuiTypography-root"
      data-testid="greetings"
    >
      Olá, Tester
    </p>
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-16nuxne-MuiButtonBase-root-MuiButton-root"
      data-testid="logoutButton"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Logout
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</header>
`;

exports[`common/components/templates/HomePageTemplate/Header unlogged matches snapshot 1`] = `
.c0 {
  height: 4rem;
  cursor: pointer;
}

<header
  className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-zhueze-MuiPaper-root-MuiAppBar-root"
  data-testid="app-bar"
>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-nb5wfo-MuiToolbar-root"
  >
    <a
      href="/"
    >
      <img
        alt="logo"
        className="c0"
        src="\\"Logo.png\\""
      />
    </a>
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
      data-testid="menuButton"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="MenuIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </button>
  </div>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-mzps2i-MuiToolbar-root"
  >
    <a
      href="/"
    >
      <img
        alt="logo desktop"
        className="c0"
        src="\\"Logo.svg\\""
      />
    </a>
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedWarning MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-4umy8a-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Fale conosco!!
      </p>
    </button>
  </div>
</header>
`;
